### DISCUSSION TITLE
How do I get the client id?

### DISCUSSION THREAD

**Discussion by CovertCode** [uNetworking/uWebSockets.js#481] [Q&A]
Let's say there are 10 clients connected to my socket server and I want to send a message to specific client from my backend. So for this I think there's need of client id or client address. But I checked the documentation there's no mention of client id of list of connected clients.

Can someone give me directions of how to achieve this? I jumped directly to uWebSocket and I'm new to this.

**Comment by hst-m** [comment]
You give it whatever ID you want, and store it to look up later however you want

store the ws by ID in a [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map):
```javascript
const wsMap=new Map()

open:ws=>wsMap.set(id,ws)
close:ws=>wsMap.delete(id)

wsMap.get(id).send(message)
```
or use [Subscribe](https://unetworking.github.io/uWebSockets.js/generated/interfaces/WebSocket.html#subscribe) / [Publish](https://unetworking.github.io/uWebSockets.js/generated/interfaces/WebSocket.html#publish) by ID:
```javascript
open:ws=>ws.subscribe(id)

app.publish(id,message)
```

**Reply by Totalus** [reply]
How does those two solutions compare memory wise ? Any idea ?

I'm wondering if uWS already maintains an internal mapping of sockets that could be used to retreive the socket instead of creating a new mapping, which would save the additional mapping memory.

**Reply by kronos-999** [reply]
A quick example with redis would be helpful and appreciated. I do not seem able to store and retrieve the ws object to use ws.send later

**Reply by e3dio** [reply]
You store references to sockets in Objects or Maps `socketsObject[id] = ws` or `socketsMap.set(id, ws)` and send `socketsObject[id].send(msg)` or `socketsMap.get(id).send(msg)` Inserting and deleting from Map is supposed to be faster than Object

**Reply by ghost** [reply]
> A quick example with redis would be helpful and appreciated. I do not seem able to store and retrieve the ws object to use ws.send later

You can't store a ws object to disk because it's not serializable and it would make no sense if it was.

**Reply by ghost** [reply]
If you're talking about some kind of session id, that's implemented as a cookie. You can get cookies like in example UpgradeSync but overall I think you're projecting past experiences on to this project. This project is somewhat less abstract.
